primitives/nomad/signature/src/signature.rs

```diff
@@ -0,0 +1,286 @@
+		};
+
+		let (recoverable_sig, _recovery_id) = self.as_signature()?;
+		let verify_key =
+			recoverable_sig.recover_verify_key_from_digest_bytes(message_hash.as_ref().into())?;
+
+		let public_key = K256PublicKey::from(&verify_key);
+		let public_key = public_key.to_encoded_point(/* compress = */ false);
+		let public_key = public_key.as_bytes();
+		debug_assert_eq!(public_key[0], 0x04);
+		let hash = crate::utils::keccak256(&public_key[1..]);
+		Ok(Address::from_slice(&hash[12..]))
+	}
+
+	/// Retrieves the recovery signature.
+	fn as_signature(&self) -> Result<(RecoverableSignature, RecoveryId), SignatureError> {
+		let recovery_id = self.recovery_id()?;
+		let signature = {
+			let mut r_bytes = [0u8; 32];
+			let mut s_bytes = [0u8; 32];
+			self.r.to_big_endian(&mut r_bytes);
+			self.s.to_big_endian(&mut s_bytes);
+			let gar: &GenericArray<u8, U32> = GenericArray::from_slice(&r_bytes);
+			let gas: &GenericArray<u8, U32> = GenericArray::from_slice(&s_bytes);
+			let sig = K256Signature::from_scalars(*gar, *gas)?;
+			RecoverableSignature::new(&sig, recovery_id)?
+		};
+
+		Ok((signature, recovery_id))
+	}
+
+	/// Retrieve the recovery ID.
+	pub fn recovery_id(&self) -> Result<RecoveryId, SignatureError> {
+		let standard_v = normalize_recovery_id(self.v);
+		Ok(RecoveryId::new(standard_v)?)
+	}
+
+	/// Copies and serializes `self` into a new `Vec` with the recovery id included
+	#[allow(clippy::wrong_self_convention)]
+	pub fn to_vec(&self) -> Vec<u8> { self.into() }
+}
+
+fn normalize_recovery_id(v: u64) -> u8 {
+	match v {
+		0 => 0,
+		1 => 1,
+		27 => 0,
+		28 => 1,
+		v if v >= 35 => ((v - 1) % 2) as _,
+		_ => 4,
+	}
+}
+
+impl<'a> TryFrom<&'a [u8]> for Signature {
+	type Error = SignatureError;
+
+	/// Parses a raw signature which is expected to be 65 bytes long where
+	/// the first 32 bytes is the `r` value, the second 32 bytes the `s` value
+	/// and the final byte is the `v` value in 'Electrum' notation.
+	fn try_from(bytes: &'a [u8]) -> Result<Self, Self::Error> {
+		if bytes.len() != 65 {
+			return Err(SignatureError::InvalidLength(bytes.len()));
+		}
+
+		let v = bytes[64];
+		let r = U256::from_big_endian(&bytes[0..32]);
+		let s = U256::from_big_endian(&bytes[32..64]);
+
+		Ok(Signature { r, s, v: v.into() })
+	}
+}
+
+impl FromStr for Signature {
+	type Err = SignatureError;
+
+	fn from_str(s: &str) -> Result<Self, Self::Err> {
+		let s = s.strip_prefix("0x").unwrap_or(s);
+		let bytes = hex::decode(s)?;
+		Signature::try_from(&bytes[..])
+	}
+}
+
+impl From<&Signature> for [u8; 65] {
+	fn from(src: &Signature) -> [u8; 65] {
+		let mut sig = [0u8; 65];
+		let mut r_bytes = [0u8; 32];
+		let mut s_bytes = [0u8; 32];
+		src.r.to_big_endian(&mut r_bytes);
+		src.s.to_big_endian(&mut s_bytes);
+		sig[..32].copy_from_slice(&r_bytes);
+		sig[32..64].copy_from_slice(&s_bytes);
+		// TODO: What if we try to serialize a signature where
+		// the `v` is not normalized?
+		sig[64] = src.v as u8;
```

Comment:
This is safe because `sig.v` can only ever be 27 or 28 here. Regarding EIP-155, this is only applied for transaction signing and this happens after the fact (after signature is created, you manually change the `v` value)

Reply:
Added comment above line to clarify